WebApi/Controllers/DataController.cs

```diff
@@ -58,34 +59,50 @@ public IHttpActionResult GetAllBarData([FromBody]UserAndBarsModel userAndBars)
                 return NotFound();
             }
             List<BarData> result = await BarFetcher.RequestBarsAroundCoords(locationRequest);
-            return Ok(result);
+            var names = result.Select(x => x.BarId).ToList();
+            var dbMgr = new DatabaseManager();
+            result = dbMgr.GetAllBarData(names);
+            dbMgr.SaveBars(result);
+            return Ok(BarDataToResponseModel(result));
         }
 
         [EnableCors(origins: "*", headers: "*", methods: "*")]
         [HttpPost]
-        public IHttpActionResult SaveBarRating([FromBody] RatingModel barObject)
+        public IHttpActionResult SaveBarRating([FromBody]RatingModel barObject)
         {
             var dbManager = new DatabaseManager();
             if (!dbManager.LogIn(barObject.User))
+            {
                 return BadRequest("User authentication failed");
-            dbManager.SaveBarRating(barObject.BarID, barObject.User.Username , barObject.Rating);
-            return Ok("Success");
+            }
+                
+            if (dbManager.UserCanVote(barObject.BarID, barObject.User, out var cooldown))
+            {
+                try
+                {
+                    dbManager.SaveBarRating(barObject.BarID, barObject.User, barObject.Rating);
+                }
+                catch (InvalidOperationException e)
+                {
+                    Console.WriteLine(e);
+                    return InternalServerError();
+                }
+                return Ok("Success");
+            }
+            return BadRequest("User cannot vote for " + cooldown.Minutes + " min. on this bar");
         }
 
         [EnableCors(origins: "*", headers: "*", methods: "*")]
         [HttpPost]
-        public IHttpActionResult GetBarRatings([FromBody]UserAndBarModel userAndBar)
-        {
-            var dbManager = new DatabaseManager();
-            var result = dbManager.GetBarRatings(userAndBar.Bar, userAndBar.User.Username);
-            return Ok(result);
-        }
-        //TEMPORARY: i think we only need 2 methods in WebApi for now: SaveBarRating(username, rating, barId) and GetBarsByCoordinates(username?, latitude, longitude, radius?)
-        [EnableCors(origins: "*", headers: "*", methods: "*")]
-        [HttpPost]
-        public async Task<IHttpActionResult> GetBarsByCoordinates([FromBody]Coordinate coordinate)
+        public async Task<IHttpActionResult> GetBarsByCoordinates([FromBody] Coordinate coordinate)
         {
-            var locationRequest = new LocationRequestModel { Latitude = coordinate.lat, Longitude = coordinate.lng, Radius = 150, User =  "test"  };
+            var locationRequest = new LocationRequestModel
+            {
+                Latitude = coordinate.lat,
+                Longitude = coordinate.lng,
+                Radius = coordinate.radius,
+                User = "test"
```

Comment:
Do we really need a user at this point?

Reply:
Thanks for pointing that out. Removed this method altogether, will use the one that uses User powers.

Or do you mean, that we don't need a username for bar fetching?